### PR TITLE
Removed unnecessary call to reversed in for loop

### PR DESCRIPTION
**Pull request by RScrusoe**
In random module's shuffle method while iterating over the list in reversed fashion, the code was using 2 generators (reversed, range). But as we know, we can simply use range generator alone (with proper start, end and step) to get the desired reversed iteration.

This is my first PR into my favourite programming language. Your feedback/suggestions are welcome and appreciated!

Note: As this is a minor change, have not opened an issue

### CODE REVIEW

**Comment by RScrusoe** [comment]
I am not able to add "skip issue" & "skip news", can the maintainers add it? or maybe I can add, but not able to find it?

File: Lib/random.py [hunk 1]
```diff
@@ -355,7 +355,8 @@ def shuffle(self, x):
         """Shuffle list x in place, and return None."""
 
         randbelow = self._randbelow
-        for i in reversed(range(1, len(x))):
+        # Iterate over the list x in reverse order
+        for i in range(len(x)-1,0,-1):
```
Comment (eendebakpt) [inline Lib/random.py]:
```suggestion
        for i in range(len(x) - 1, 0, -1):
```
Follow the coding conventions, see https://peps.python.org/pep-0008/

**Comment by eendebakpt** [comment]
It original code indeed used two generators, and this PR only one. But is this PR more readable (I think not) or faster (benchmark)?

**Comment by RScrusoe** [comment]
@eendebakpt Thank you for the review. Fixed it!

For the comment on readability, I really believe, that people (who read this code) would know the range iterator and how it is used with start/end/step params. 

On the benchmark: I am not sure on how to benchmark this particular small thing, as there can be too many things that would interfere with benchmarking code which would try to benchmark this particular difference (at least, I am not so sure of how to perf test it). But, looking at proposed code, **it is clear that an additional function call (stack frame allocation) is saved.**

**Comment by eendebakpt** [comment]
You can do a quick benchmark (using ipython) with 
```
import random
x = list(range(100))
%timeit random.shuffle(x)
```
That does not tell everything, but is a first indication of whether the code is faster or not.

You do save a function call, but in the pr the arguments to range are more complex, so the end result could be faster, slower or ... about the same in performance.

**Comment by TeamSpen210** [comment]
`range` does [have a `__reversed__`](https://github.com/python/cpython/blob/2bb7ab7ad364ec804eab8ed6867df01ece887240/Objects/rangeobject.c#L1182) method, meaning both versions will produce the same iterator object. So the question is whether the Python-level arithmetic is faster than calling `reversed()`.

**Comment by eendebakpt** [comment]
There was a detailed discussion in https://github.com/python/cpython/issues/108598 on improving the performance of `random.shuffle`. 

@pochmann Is there a reason why for the proposal3b from https://github.com/python/cpython/issues/108598#issuecomment-1702500422 no PR was created?

**Comment by AA-Turner** [comment]
I suspect this is unlikely to be worth it, absent incredible microbenchmarks.

A

**Comment by RScrusoe** [comment]
Sorry for late reply!
PFB my attempt at benchmarks
I spawned a fresh EC2 instance (c5.large) with ubuntu 22.04
Code that I ran:
```
import random
import timeit

def test_shuffle():
    # Generate a small list
    small_list = list(range(10))
    random.shuffle(small_list)

if __name__ == "__main__":
    # Benchmark the shuffle function over a large number of iterations
    iterations = 1000000  # Number of times to run the shuffle
    time_taken = timeit.timeit("test_shuffle()", setup="from __main__ import test_shuffle", number=iterations)

    print(f"Time taken to shuffle small lists {iterations} times: {time_taken:.4f} seconds")
```

I specifically tested small lists but large number of times shuffling call.


My benchmark finding:
Original:
Screenshot 2024-10-21 at 11 17 29 AM

The PR version:
Screenshot 2024-10-21 at 11 15 58 AM

Math:
Avg of 3 runs with original code: 2.5849
Avg of 3 runs with PR code:         2.5240
The PR code is better by %:         **2.35%**
(this is for 1 million calls to the shuffle function)


Suggestions/feedback is welcome

**Comment by rhettinger** [comment]
The code uses `reversed` because it is more clear than the `range` variation. See [`PEP 322`](https://peps.python.org/pep-0322/)

Note the `range` object has custom C code implementing reverse iteration. So it is quite speedy.

**Comment by RScrusoe** [comment]
@rhettinger Thanks for the feedback, makes sense!

**Pull request closed without merging**